### PR TITLE
[MPS] Faster reductions (1/5): skip input up-casts and vec4 full reductions

### PR DESCRIPTION
**Pull request by Isalia20**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* #191100
* #191099
* #191098
* #191097
* __->__ #191101
* #191104 

Two changes to how MPS reductions handle dtypes and full (scalar-output) reductions:

`make_reduction` casts the input to the accumulation dtype when it differs from the output dtype, which on MPS materializes a 2-8x larger copy plus a cast kernel before the reduction starts. The Metal kernels never needed it: they are instantiated per (input, output) dtype pair and widen in-register. This PR passes the original input dtype through for integral -> int64 (new, MPS-only) and half/bf16 -> fp32, on the sum/mean overload only; the overload shared with `vector_norm`/nansum keeps casting since MPS has no (half, float) norm kernels.

Second, pass 1 of the two-pass full reduction (at most 512 contiguous slices, one threadgroup each) was walking the generic `NormParams` indexing helper one scalar at a time, even though the partition is just [num_groups, elems_per_group] over a contiguous buffer. `sum_reduction_flat` (sum/nansum/mean/count_nonzero) and `value_reduction_flat` (min/max/any/all) replace it with flat indexing, chained accumulators, and an unrolled 4-elements-per-step inner loop. Loads stay scalar, so every (op, input, output) pair with a base kernel gets a flat variant and dispatch is just strided vs flat.

<details>
<summary> 
Speedup over main
</summary>

| shape | dim | case | before us | after us | speedup |
| --- | --- | --- | ---: | ---: | ---: |
| `(1024,)` | full | f32 sum | 4.5 | 4.7 | 0.96x |
| `(1024,)` | full | bf16 sum | 3.7 | 3.7 | 1.01x |
| `(1024,)` | full | bf16 amax | 4.1 | 3.9 | 1.05x |
| `(1024,)` | full | i32 sum | 6.0 | 4.1 | 1.46x |
| `(1024,)` | full | bf16 sum, fp32 out | 5.6 | 3.5 | 1.60x |
| `(4096,)` | full | f32 sum | 3.8 | 3.9 | 0.99x |
| `(4096,)` | full | bf16 sum | 3.8 | 3.8 | 0.99x |
| `(4096,)` | full | bf16 amax | 4.1 | 4.2 | 0.96x |
| `(4096,)` | full | i32 sum | 6.3 | 4.1 | 1.55x |
| `(4096,)` | full | bf16 sum, fp32 out | 6.0 | 3.8 | 1.55x |
| `(8192,)` | full | f32 sum | 4.2 | 4.4 | 0.95x |
| `(8192,)` | full | bf16 sum | 4.1 | 3.8 | 1.07x |
| `(8192,)` | full | bf16 amax | 4.5 | 4.7 | 0.96x |
| `(8192,)` | full | i32 sum | 7.0 | 4.5 | 1.55x |
| `(8192,)` | full | bf16 sum, fp32 out | 6.1 | 4.1 | 1.51x |
| `(10000,)` | full | f32 sum | 6.3 | 5.5 | 1.14x |
| `(10000,)` | full | bf16 sum | 6.2 | 5.8 | 1.08x |
| `(10000,)` | full | bf16 amax | 6.9 | 6.2 | 1.12x |
| `(10000,)` | full | i32 sum | 9.4 | 6.3 | 1.49x |
| `(10000,)` | full | bf16 sum, fp32 out | 8.5 | 5.7 | 1.49x |
| `(16384,)` | full | f32 sum | 6.7 | 5.6 | 1.21x |
| `(16384,)` | full | bf16 sum | 6.4 | 5.6 | 1.15x |
| `(16384,)` | full | bf16 amax | 7.2 | 6.1 | 1.17x |
| `(16384,)` | full | i32 sum | 9.7 | 6.2 | 1.58x |
| `(16384,)` | full | bf16 sum, fp32 out | 8.9 | 5.5 | 1.63x |
| `(32768,)` | full | f32 sum | 6.7 | 5.7 | 1.19x |
| `(32768,)` | full | bf16 sum | 6.5 | 5.5 | 1.18x |
| `(32768,)` | full | bf16 amax | 7.3 | 6.4 | 1.14x |
| `(32768,)` | full | i32 sum | 10.3 | 6.2 | 1.66x |
| `(32768,)` | full | bf16 sum, fp32 out | 9.3 | 5.7 | 1.64x |
| `(65536,)` | full | f32 sum | 6.9 | 5.7 | 1.22x |
| `(65536,)` | full | bf16 sum | 6.5 | 5.3 | 1.23x |
| `(65536,)` | full | bf16 amax | 7.2 | 6.2 | 1.17x |
| `(65536,)` | full | i32 sum | 10.4 | 6.0 | 1.73x |
| `(65536,)` | full | bf16 sum, fp32 out | 9.4 | 5.6 | 1.69x |
| `(131072,)` | full | f32 sum | 6.7 | 5.8 | 1.16x |
| `(131072,)` | full | bf16 sum | 6.3 | 5.3 | 1.20x |
| `(131072,)` | full | bf16 amax | 7.0 | 6.0 | 1.17x |
| `(131072,)` | full | i32 sum | 11.1 | 6.2 | 1.78x |
| `(131072,)` | full | bf16 sum, fp32 out | 11.0 | 5.6 | 1.95x |
| `(262144,)` | full | f32 sum | 7.7 | 6.3 | 1.23x |
| `(262144,)` | full | bf16 sum | 7.2 | 5.8 | 1.23x |
| `(262144,)` | full | bf16 amax | 8.5 | 6.8 | 1.25x |
| `(262144,)` | full | i32 sum | 14.2 | 6.7 | 2.13x |
| `(262144,)` | full | bf16 sum, fp32 out | 14.5 | 5.8 | 2.52x |
| `(524288,)` | full | f32 sum | 9.7 | 7.1 | 1.37x |
| `(524288,)` | full | bf16 sum | 9.2 | 6.1 | 1.50x |
| `(524288,)` | full | bf16 amax | 11.6 | 7.8 | 1.48x |
| `(524288,)` | full | i32 sum | 29.5 | 7.8 | 3.77x |
| `(524288,)` | full | bf16 sum, fp32 out | 20.9 | 6.3 | 3.31x |
| `(1048576,)` | full | f32 sum | 15.9 | 11.7 | 1.36x |
| `(1048576,)` | full | bf16 sum | 12.6 | 7.3 | 1.73x |
| `(1048576,)` | full | bf16 amax | 17.1 | 10.3 | 1.67x |
| `(1048576,)` | full | i32 sum | 52.9 | 12.2 | 4.36x |
| `(1048576,)` | full | bf16 sum, fp32 out | 33.7 | 7.3 | 4.63x |
| `(4194304,)` | full | f32 sum | 48.1 | 37.4 | 1.29x |
| `(4194304,)` | full | bf16 sum | 34.0 | 21.5 | 1.58x |
| `(4194304,)` | full | bf16 amax | 51.3 | 34.5 | 1.49x |
| `(4194304,)` | full | i32 sum | 307.7 | 37.8 | 8.13x |
| `(4194304,)` | full | bf16 sum, fp32 out | 114.3 | 21.6 | 5.29x |
| `(16777216,)` | full | f32 sum | 243.0 | 233.4 | 1.04x |
| `(16777216,)` | full | bf16 sum | 118.2 | 108.2 | 1.09x |
| `(16777216,)` | full | bf16 amax | 116.7 | 108.1 | 1.08x |
| `(16777216,)` | full | i32 sum | 1266.9 | 233.9 | 5.42x |
| `(16777216,)` | full | bf16 sum, fp32 out | 644.7 | 108.2 | 5.96x |
| `(2048, 2048)` | 1 | i32 sum | 302.3 | 35.9 | 8.43x |
| `(2048, 2048)` | 0 | i32 sum | 298.3 | 37.4 | 7.98x |
| `(2048, 2048)` | full | i32 sum | 308.2 | 37.9 | 8.12x |
| `(2048, 2048)` | 1 | bf16 sum, fp32 out | 102.9 | 20.2 | 5.10x |
| `(2048, 2048)` | 1 | f16 sum, fp32 out | 101.7 | 20.1 | 5.05x |
| `(2048, 2048)` | full | bf16 sum | 34.2 | 21.5 | 1.59x |
| `(1048576,)` | full | i64 amax | 25.8 | 22.9 | 1.13x |
| `(32,)` | full | f32 sum (enqueue floor) | 3.4 | 3.4 | 0.99x |

</details>

<details> 
<summary> 
Benchmark script:
</summary>

```python
import csv
import sys
import time

import torch

WARMUP = 200
ITERS = 200
ROUNDS = 7

SIZES = (
    1024,
    4096,
    8192,
    10000,
    16384,
    32768,
    65536,
    131072,
    262144,
    524288,
    1 << 20,
    1 << 22,
    1 << 24,
)

VARIANTS = (
    ("f32 sum", torch.float32, lambda x: x.sum()),
    ("bf16 sum", torch.bfloat16, lambda x: x.sum()),
    ("i32 sum", torch.int32, lambda x: x.sum()),
    ("bf16 sum, fp32 out", torch.bfloat16, lambda x: x.sum(dtype=torch.float32)),
    ("bf16 amax", torch.bfloat16, lambda x: x.amax()),
)

CASES = [
    (f"({n},)", "full", case, dtype, (n,), fn) for n in SIZES for case, dtype, fn in VARIANTS
] + [
    ("(2048, 2048)", "1", "i32 sum", torch.int32, (2048, 2048), lambda x: x.sum(1)),
    ("(2048, 2048)", "0", "i32 sum", torch.int32, (2048, 2048), lambda x: x.sum(0)),
    ("(2048, 2048)", "full", "i32 sum", torch.int32, (2048, 2048), lambda x: x.sum()),
    ("(2048, 2048)", "1", "bf16 sum, fp32 out", torch.bfloat16, (2048, 2048), lambda x: x.sum(1, dtype=torch.float32)),
    ("(2048, 2048)", "1", "f16 sum, fp32 out", torch.float16, (2048, 2048), lambda x: x.sum(1, dtype=torch.float32)),
    ("(2048, 2048)", "full", "bf16 sum", torch.bfloat16, (2048, 2048), lambda x: x.sum()),
    ("(1048576,)", "full", "i64 amax", torch.int64, (1 << 20,), lambda x: x.amax()),
    ("(32,)", "full", "f32 sum (enqueue floor)", torch.float32, (32,), lambda x: x.sum()),
]


def make(dtype, shape):
    if dtype.is_floating_point:
        return torch.randn(shape, device="mps", dtype=dtype)
    return torch.randint(-8, 8, shape, device="mps", dtype=dtype)


def bench(fn, x):
    for _ in range(WARMUP):
        fn(x)
    torch.mps.synchronize()
    best = float("inf")
    for _ in range(ROUNDS):
        t0 = time.perf_counter()
        for _ in range(ITERS):
            fn(x)
        torch.mps.synchronize()
        best = min(best, (time.perf_counter() - t0) / ITERS * 1e6)
    return best


def main():
    torch.manual_seed(0)
    with open(sys.argv[1], "w", newline="") as f:
        w = csv.writer(f)
        w.writerow(("shape", "dim", "case", "us", "git_version"))
        for shape_name, dim, case, dtype, shape, fn in CASES:
            x = make(dtype, shape)
            us = bench(fn, x)
            w.writerow((shape_name, dim, case, f"{us:.6f}", torch.version.git_version))
            print(f"{shape_name:<14} {dim:<5} {case:<24} {us:8.1f} us", flush=True)


if __name__ == "__main__":
    main()

```

</details>

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/ReduceOps.mm [hunk 1]
```diff
@@ -980,8 +980,15 @@ static void reduction_dispatch_mps(TensorIterator& iter, const ReductionDispatch
       const uint32_t elems_per_group = reduction_size / num_groups;
       auto partials = at::empty({(int64_t)num_groups}, output.options().dtype(opts.partial_dtype));
 
-      auto p1_kernel =
-          fmt::format("{}reduction{}_{}_{}", opts.prefix, use_strided ? "_strided" : "", in_str, partial_str);
+      const bool vec_ok = opts.input_kernel_dtype == kFloat || opts.input_kernel_dtype == kHalf ||
+          opts.input_kernel_dtype == kBFloat16 || opts.input_kernel_dtype == kInt || opts.input_kernel_dtype == kLong ||
+          opts.input_kernel_dtype == kShort;
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/ReduceOps.mm]: Any comment as to why?

File: aten/src/ATen/native/ReduceOpsUtils.h [hunk 2]
```diff
@@ -219,7 +219,11 @@ inline TensorIterator make_reduction(
   // product of templated kernel launches.
   const bool gpu_lowp_to_f32 = (
         (self.is_cuda() || self.is_xpu()) && (self.scalar_type() == kHalf || self.scalar_type() == kBFloat16) && out_dtype == kFloat);
```
Comment (malfet) [inline aten/src/ATen/native/ReduceOpsUtils.h:221]: Why not add MPS here?

File: aten/src/ATen/native/mps/operations/ReduceOps.mm [hunk 3]
```diff
@@ -980,8 +980,15 @@ static void reduction_dispatch_mps(TensorIterator& iter, const ReductionDispatch
...
+      const bool vec_op = opts.prefix == "sum_" || opts.prefix == "nansum_" || opts.prefix == "min_" ||
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/ReduceOps.mm]: Same as above, any comment as to why?

File: aten/src/ATen/native/mps/kernels/ReduceOps.metal [hunk 4]
```diff
@@ -904,6 +992,63 @@ kernel void value_reduction(
...
+  }
+  const uint stride = tptg * NCHAINS;
+  uint i = tid;
+  for (; i + (NCHAINS - 1) * tptg < n_vec; i += stride) {
+    for (uint j = 0; j < NCHAINS; j++) {
+      const V v = vin[i + j * tptg];
+      acc[j] = Op::combine(acc[j], Load::template load<TA>(v.x));
+      acc[j] = Op::combine(acc[j], Load::template load<TA>(v.y));
+      acc[j] = Op::combine(acc[j], Load::template load<TA>(v.z));
+      acc[j] = Op::combine(acc[j], Load::template load<TA>(v.w));
```
Comment (malfet) [inline aten/src/ATen/native/mps/kernels/ReduceOps.metal]:
```suggestion
    #pragma unroll loop
     for(int k = 0; k < 4; ++k) {
      acc[j] = Op::combine(acc[j], Load::template load<TA>(v[k]));
    }
```

**Review by malfet — CHANGES_REQUESTED** [review]
Mostly sounds reasonable, but a few comments:
- In ReductionUtils, why not include MPS in upcast logic (or may be move all of that upcast logic into a separate PR)
- `vec_ok` and `vec_op`  sounds just a bit too similar, and it's unclear what each one should mean, and why those dtypes and not other. 
Hardcoding some string matching in dispatcher itself feels wrong, imo it should be a callee option arg ( say `supports_ilp` because it's not really vectorized operations that are faster, it's the same logic that doing multiple reduction per thread is better than trying to reduce right away. And I might even advice to stay away from vectorized types and just do small array)

File: aten/src/ATen/native/mps/operations/ReduceOps.mm (hunk 1 above)
Comment (Isalia20) [reply aten/src/ATen/native/mps/operations/ReduceOps.mm]: This list just tracked which _vec kernels compiled/were registered (vec4type_t types only covered 6 dtypes)

File: aten/src/ATen/native/mps/operations/ReduceOps.mm (hunk 3 above)
Comment (Isalia20) [reply aten/src/ATen/native/mps/operations/ReduceOps.mm]: This tracked which kernels existed, the only missing one was `count_nonzero` because that kernel would accumulate it in a wrong way, `count_nonzero` sums 1s and counts in uint (i.e. we know how many elements tensor have so can use specialized kernel even for int64 tensors if we know that the numel won't overflow uint32)

The vec kernel hardcoded its accumulator to `opmath_t<TO> = int64`. Rather than add that special case, didn't register `count_nonzero` kernels.

File: aten/src/ATen/native/ReduceOpsUtils.h (hunk 2 above)
Comment (Isalia20) [reply aten/src/ATen/native/ReduceOpsUtils.h:221]: That overload also feeds `vector_norm`, `powsum` and `std/var`, for which MPS kernels are only instantiated for matching input/output dtypes, I think this change can go into a followup PR

File: aten/src/ATen/native/mps/kernels/ReduceOps.metal (hunk 4 above)
Comment (Isalia20) [reply aten/src/ATen/native/mps/kernels/ReduceOps.metal]: done

File: aten/src/ATen/native/mps/operations/ReduceOps.mm (hunk 1 above)
Comment (Isalia20) [reply aten/src/ATen/native/mps/operations/ReduceOps.mm]: Deleted in the update since scalar loads no longer require this

File: aten/src/ATen/native/mps/operations/ReduceOps.mm (hunk 3 above)
Comment (Isalia20) [reply aten/src/ATen/native/mps/operations/ReduceOps.mm]: Updated, all ops have the flat path now (added it for count_nonzero as well)

**Review by malfet — APPROVED** [review]
Thank you for the update, LGTM
